### ISSUE TITLE
Bug: [prefer-readonly] autofixer doesn't add type to property that is mutated in the constructor

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#1056]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly": 2
  }
}
```

```TS
// your repro code case
class X {
  private _isValid? = true;
  get isValid(): boolean {
    return this._isValid;
  }

  constructor(data?: {}) {
    if (!data) {
      this._isValid = false;
    }
  }
}
```

**Expected Result**

No report

**Actual Result**

```sh
@typescript-eslint/prefer-readonly

Member '_isValid' is never reassigned; mark it as `readonly`
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.3.3` |
| `@typescript-eslint/parser`        | `2.3.3` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.5.1` |
| `node`                             | `12.11.1` |
| `yarn`                             | `1.19.0` |

**Comment by bradzacher** [comment]
This isn't actually wrong. `_isValid` is never reassigned.
I think that you're misunderstanding how classes work in JS.

Assigning a property a default value at declaration time is exactly the same as assigning a default value in the constructor:

```ts
class X {
  myVar: number = 1;
}
// is equivalent to
class X {
  myVar: number;
  constructor() {
    this.myVar = 1;
  }
}
```

This means that your example is actually the same as
```
class X {
  private _isValid: boolean;

  constructor(data?: {}) {
    this._isValid = true;
    if (!data) {
      this._isValid = false;
    }
  }
}
```

As far as TS is concerned - a property is not readonly during the constructor.
I.e. your code is correct if you add readonly:
```ts
class X {
  private readonly _isValid: boolean = true;
  get isValid(): boolean {
    return this._isValid;
  }

  constructor(data?: {}) {
    if (!data) {
      this._isValid = false;
    }
  }
}
```
[repl](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=13&pc=2#code/PTAEE8HsFcCdVgUwA60qAxpAJozBDAZ0QCgMAbIw0ADVAG8TRRUBLAN3wBc8l9tIAO3LhQAfVaEAavnKtsALlAAjSJHKJ8g0AF5QXWNEQBuJqADmiLqEky52ABQBKJavWbtjZsyRc42rgALSQA6CWlZeVNmAF8SMyxBQgNoDC5IWAdsbnwAfiV6GKcGM2ZWADNQBwBCbK58Yq9vZiDQ8Lt5XVBy2WJo7zjYkhigA)

**Comment by JounQin** [comment]
Thanks @bradzacher, the previous seems wrong when I'm using `private readonly _isValid = true;`, it will emit `typeof _isValid === true`, so ts will complain on `this._isValid = false;`. Change it to `private readonly _isValid: boolean = true;` manually really works, but Should it be handled by `eslint-plugin` autofixing?

**Comment by bradzacher** [comment]
could potentially, but it'll be a really hard fix to implement.
You'll have to detect the type of the value, and print that as a type annotation on the property.

Happy to accept a PR for it. I'll reopen this in case someone wants to implement it.

**Comment by JoshuaKGoldberg** [comment]
Note that this only happens for members, not private parameter properties.

```ts
class X {
  private _isValid = true; // fix suggests readonly, but should also suggest : boolean

  getIsValid = () => this._isValid;

  constructor(data?: {}) {
    if (!data) {
      this._isValid = false;
    }
  }
}

class Y {
  getIsValid = () => this._isValid;

  constructor(private readonly _isValid = true) { // no issues
    if (Math.random() > 0.5) {
      this._isValid = false;
    }
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
Aha, yet another issue with an ideal resolution blocked on TypeScript exposing an assignability API: https://github.com/microsoft/TypeScript/issues/9879

In order to know whether the fix should add a type annotation, we'd need to check the initializer value against every possible assigned value in the constructor. This holds for all primitives and even unions of complex objects. [playground link with more unions](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&showAST=es&code=MYGwhgzhAEAaBCB7RICmYB20DeBYAUNEdAA4BOAlgG5gAuq0Z6AJohiAJ7QD6FEAamBAVm0ALzRaZAK6oA3AQLFoAc1S0AkgKEjx0ABQBKcQD5JACz4A6XtuHMF+JcWBsIU6cFqIy+5nTAAfgAuHABfYzxCZWIKADMDf1owSOcYmNpLCBs+QXs9OKEIeTSYsNLy-EqCUEgYWABlKQoMFRxS8mo6BiYwVnYuWzzdCQByUcdStU07EYNjMTNM6yGdB0VoolcMdxkvHz8AkPDUzZj4xIDT9PTl7NX8sfMwCdLlSveCavxaqDgAeQARgArVBeGBRZSdGj0RgsNicHi5NZ6ADao0sowAuhtlNMtMNRBIjKYLCtkfZJmdtrtPN5fEkgqFsBF2mdlBdDslrjdlHccrMiThoDQQLJQh4GB8btKiJVKkA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9DCnAGZwpbkCGAJgPYG5QgBcI1KTKIAXyA&tsConfig=N4XyA)

```ts
class XObjects {
    private readonly _isValid = ['hi']

    getIsValid = () => this._isValid;

    constructor(data?: {}) {
        if (data) {
            this._isValid = { value: true }
        }
    }
}
```

Marking as blocked on TypeScript's type relationship API.

**Comment by JoshuaKGoldberg** [comment]
Now that v8 is released, per https://github.com/typescript-eslint/typescript-eslint/discussions/7936, we can mark this issue as unblocked! 🚀

**Comment by ronami** [comment]
Interesting!

Writing some of my thoughts after digging into the issue:

To my understanding, the problem is closely related to [TypeScript's type-widening literal types](https://github.com/microsoft/TypeScript-New-Handbook/blob/master/reference/Widening-and-Narrowing.md#literal-widening). It seems that when a member (without an explicit type annotation) is assigned a default value, TypeScript infers its type based on the type of the initializer, but also applies type-widening if the initializer is a literal type:

```ts
class X {
  // `a' is inferred as `boolean` and not as `true`
  private a = true;
}
```

When a `readonly` modifier is applied to that member, TypeScript doesn't widen the type:

```ts
class X {
  // `a' is inferred as `true` and not as `boolean`
  private readonly a = true;
}
```

This is similar to assigning a variable with `const` or `let`, or using `as const`. Also, it seems that widening literals only happens to literal types that originate from expressions, not types ([more on this here](https://github.com/microsoft/TypeScript-New-Handbook/blob/master/reference/Widening-and-Narrowing.md#fresh-literal-types-and-mutable-locations)):

```ts
declare const a: 'foo';

// `b` has the non-widened type of 'foo'
let b = a;

// `c` has the widened type of `string`
let c = 'foo';
```

---

Taking that into consideration, I think the rule should only add a type annotation as part of its auto-fix if the member has an initializer and doesn't have a type annotation. If there is a type annotation, there's nothing to add.

If there's no initializer, it seems that the type is inferred by TypeScript, based on assignments on the constructor, regardless of whether the member is marked as `readonly` or not:

```ts
class X {
  // `a` is inferred as `string | number | boolean` or as `any` of `strict` isn’t turned on
  private a;

  constructor(data?: unknown) {
    this.a = 'true';

    if (!data) {
      this.a = false;
    }

    if (typeof data === 'string') {
      this.a = 10;
    }
  }
}
```

If `strict` mode is off, TypeScript would infer `a` as being implicit `any`, regardless if there's a `readonly` modifier or not.

Assuming there's always an initializer and no type annotation, in my opinion, the only time when a type annotation should be added is when type widening will change the type of the member (and then the annotation preserves the original type).

This means that the rule shouldn't fix pre-existing type errors, like in the following case:

```ts
class X {
  private a = ['foo']

  constructor(data?: unknown) {
    if (data) {
      // type error with or without the `readonly` modifier, as the inferred type of `a` is `string[]`
      this.a = { value: true }
    }
  }
}
```

I think the rule should only prevent type errors that may be caused by the addition of a `readonly` modifier (which prevents literal type narrowing).

The fixer would also need to add the correct type annotation. Since the rule only has to add a type annotation for literal types (like `number`, `boolean`, `string`), I think `checker.typeToString()` can work out as a way to generate the fixer's type annotation.

The enum type is one potentially more complex type for which the rule will need to add a type annotation. Testing this in some edge cases, there are cases where `readonly` will change the type of the expression, but adding the relevant type annotation is complex:

```ts
// foo.ts
enum Foo {
  Bar,
  Bazz,
}

export const bar = Foo.Bar;

// index.ts
import { bar } from './foo';

export class Test {
  private foo = bar;
}
```

In the example above, adding the `: Foo` type annotation is correct (as `readonly` will change the type from `Foo` to `Foo.Bar`), but `Foo` isn't available in the scope of the violating member. I can think of ways around this, but none are optimal.
